### ISSUE TITLE
Dijkstra's Algorithm - Consider Whether Tiles have Roads

### ISSUE THREAD

**Issue by QuintillusCFC** [C7-Game/Prototype#191]
Depends on #100 .

Road tiles will only cost 1/3 of a movement point (by default; this should be configurable).  This affects their weight in terms of unit pathing.

This could also include that rivers _don't_ get that bonus (without a future bridge-building tech), but that can be spun off into a separate issue if this doesn't turn out to be easy.

**Comment by Kright** [comment]
We use integers for movement costs, but road requires fractional parts (like 1/3). In addition, there are railroads. For example, in 4th civ it has movement cost 1/10.  
What we should do with this? Use floats and be careful because of possible rounding errors?

**Comment by QuintillusCFC** [comment]
That is a great question and one where we haven't decided on an answer.  As you say, floats could be used, but we run the risk of three movements on road resulting in a unit having only used, e.g. 0.9999982 of its movement points, and thus being able to move again when it shouldn't (not sure if 1/3 will actually be under or over the real value, but you get the point).

Two ideas that I've thought of:

 - Use floats, but add a very small amount to each movement cost to each move that isn't free (e.g. 1/1000th of a point or less to everything that isn't like railroads in Civ3) to account for this.  In the UI, we'll only display to a certain precision, but ensuring that 1/3 costs is 0.3333334 instead of 0.3333332 can avoid errors.
 - Use fractions, numerator and denominator.  Have our own math system for movement.  A Knight starts with 2/1 movement, if it moves on a road that costs 1/3 of a movement, the math system will expand 2/1 to 6/3, and then subtract 1/3 so it now has 5/3 movement left.  If it then moves on a railroad that costs 1/10 of a movement, the 5/3 becomes 50/30, and then we subtract to get 47/30 movement points left.  In the UI, it can still be displayed as decimal, though we could print out the fractions for debugging.

I suspect the floats-but-add-a-bit would be faster and may be simpler.  But the part of me that remembers when math was my favorite subject likes the fraction idea.

There may be other good ideas I haven't though of too.

**Comment by Kright** [comment]
I think it's better to use floats and treat them as equal when abs(a - b) < eps. Eps is something small, 0.0001 for example. I'm going to make struct wrapping float and properly implement =, >, < operators for it. It will be less error-prone.

**Comment by QuintillusCFC** [comment]
That also sounds like a good idea.  I've done something similar with timestamps in the past.
